app/javascript/mastodon/actions/compose.js

```diff
@@ -212,17 +214,43 @@ export function clearComposeSuggestions() {
 };
 
 export function fetchComposeSuggestions(token) {
-  return (dispatch, getState) => {
-    api(getState).get('/api/v1/accounts/search', {
-      params: {
-        q: token,
-        resolve: false,
-        limit: 4,
-      },
-    }).then(response => {
-      dispatch(readyComposeSuggestions(token, response.data));
-    });
-  };
+  let leading = token[0];
+
+  if (leading === '@') {
+    // handle search
+    return (dispatch, getState) => {
+      api(getState).get('/api/v1/accounts/search', {
+        params: {
+          q: token.slice(1), // remove the '@'
+          resolve: false,
+          limit: 4,
+        },
+      }).then(response => {
+        dispatch(readyComposeSuggestions(token, response.data));
+      });
+    };
+  } else if (leading === ':') {
+    // shortcode
+    let allShortcodes = Object.keys(emojione.emojioneList);
+    // TODO when we have custom emojons merged, add them to this shortcode list
+    return (dispatch) => {
+      dispatch(readyComposeSuggestionsTxt(token, allShortcodes.filter((sc) => {
+        return sc.indexOf(token) === 0;
+      })));
```

Comment:
should have a take(4) or similar—no use finding all 28 matches for "clock" when we're only going to show a handful

Reply:
we actually show all of them ;D
![screenshot_20170920_214508](https://user-images.githubusercontent.com/2041118/30664129-fc9f99aa-9e4c-11e7-958d-8bbf45892cec.png)

Reply:
I can add a scrollbar to it...